docs/DesignDocs/SGX2-EDMM.md

```diff
@@ -0,0 +1,105 @@
+acknowledges its addition by issuing an `ENCLU[EACCEPT]` instruction for the
+respective page address. Further, the `ENCLS[EMODPR]` and `ENCLU[EMODPE]`
+instructions restrict or extend page permissions, respectively. Page type
+changes (including page removal/trimming) are performed by the `ENCLS[EMODT]`
+instruction.
+
+## Heap allocation
+
+We propose to add a new OCALL to OE enclaves which requests new memory pages
+from the SGX driver (via `ioctl`), combined with an enclave function that, upon
+return from the OCALL, accepts the new pages. This does not require handling of
+signals or page faults and is easily integrated with memory managers like the
+default dlmalloc, by pre-mapping a configurable maximum number of pages and
+tracking the number of currently mapped pages. For performance reasons, we
+propose to add a switchless version of the OCALL whenever
+`OE_ENCLAVE_SETTING_CONTEXT_SWITCHLESS` is enabled:
+
+    untrusted {
+      public oe_result_t oe_add_pages(size_t count);
+      public oe_result_t oe_add_pages_switchless(size_t count) transition_using_threads;
+    };
+
+The pluggable allocator API then uses `oe_add_pages`, before allocating
+additional pages, if necessary, before calling into `oe_allocator_*alloc`. The
+maximum number of heap pages (and perhaps other parameters like page
+permissions) are passed to `oe_allocator_init`.
+
+Applications that run a full OS kernel, for instance SGX-LKL, need to know about
+the size and layout of their address space and can therefore effectively use
+dynamic region allocation, which the SGX driver supports by tracking memory
+regions in the SGX driver. It automatically grows the heap by allocating a
+configurable number of pages upon request for a particular (unmapped) page.
+Dynamic regions are most conveniently declared as an initialization parameter to
+`oe_allocator_init` again.
+
+The permissions (R/W/X) of pages can be modified dynamically. Permissions of
+mapped pages are restricted by the `ENCLS[EMODPR]` instruction, which is a
+privileged instruction (ring 0) that the host executes. Similarly, permissions
+are extended by the `ENCLU[EMODPE]` instruction, which the enclave executes.
+OE's current pluggable allocators do not support permission changes. We propose
+to add a single, easy to use function to change (both restrict and extend) the
+page types of allocated memory, e.g.
+
+    oe_result_t oe_sgx_set_page_permissions(void* address, bool R, bool W, bool X);
+
+We expect that OE applications will usually want the R bit enabled by default,
+but some applications require the X bit to be set and others may chose to
+relinquish the W flag to avoid accidental modification of memory pages. (One of
+the reasons that SGX-LKL is forced to use a custom version of OE, is that the
+memory permissions are not configurable at signing, launching, or run time; see
+also [#3561](https://github.com/openenclave/openenclave/pull/3561).)
+
```

Comment:
@wintersteiger could you elaborate on call flow in SGX-LKL for permission changes? Will it load code into a heap area, then change permissions? The referred PR is closed and you stated it won't be an issue with EDMM.
I'd like to confirm the APIs proposed in #3991 for enclave side would be sufficient.

Reply:
In the past, SGX-LKL simply added the X flag to the entire heap at enclave build time and at startup. This was done via fork of OE which had the extra `SGX_SECINFO_X` added to the pages set up in `_add_filled_pages` in `create.c`, so whatever it loaded into that memory was also executable. This is not necessary with EDMM because page permissions can be changed after enclave startup, as indeed the API in #3991 includes. I'm not sure whether there will be an EDMM version of SGX-LKL, but if so, it would be a trivial change, so I think we're fine there.